tests/test_testclient.py

```diff
@@ -113,3 +113,55 @@ def app(scope):
     with client.websocket_connect("/") as websocket:
         data = websocket.receive_json()
         assert data == {"message": "test"}
+
+
+@pytest.mark.parametrize("use_middleware", (0, 1, 2))
+@pytest.mark.parametrize("raise_server_exceptions", (True, False))
+def test_error_with_middleware_and_testclient_exit(
+    use_middleware, raise_server_exceptions
+):
+    """TestClient's __exit__ should not raise the exception again."""
+    app = Starlette()
+
+    events = []
+
+    @app.on_event("startup")
+    async def startup():
+        events.append("startup")
+
+    @app.on_event("shutdown")
+    async def shutdown():
+        events.append("shutdown")
+
+    if use_middleware:
+
+        @app.middleware("http")
+        async def http_middleware(request, call_next):
+            return await call_next(request)
+
+        if use_middleware == 2:
+
+            @app.middleware("http")
+            async def second_middleware(request, call_next):
+                return await call_next(request)
+
+    if raise_server_exceptions:
+
+        @app.route("/error")
+        async def error(request):
+            pytest.fail("should_only_fail_once")
+
+        with TestClient(app, raise_server_exceptions=True) as client:
+            with pytest.raises(pytest.fail.Exception):
+                client.get("/error")
+    else:
+
+        @app.route("/error")
+        async def error(request):
+            pytest.fail("should_not_raise")
```

Comment:
Could be that exception raised by `pytest.fail` handles differently by pytest itself?

If I replace all `pytest.fail` with `raise Exception` and rollback changes in `testclient.py` the test passes fine.

Reply:
I think it is rather due to it being not of type Exception, but BaseException.

Reply:
IIRC BaseException/Exception is handled differently also in asyncio already in some place(s).

Reply:
> I think it is rather due to it being not of type Exception, but BaseException.

Indeed

Reply:
Might be related to this, but do not remember really: https://github.com/python/cpython/blob/443fe5a52a3d6a101795380227ced38b4b5e0a8b/Lib/asyncio/base_events.py#L161-L168